File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StringReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int loopCount = Integer.parseInt(br.readLine());
        // loop scope
        for(int i = 0; i < loopCount; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int docCount = Integer.parseInt(st.nextToken());
            int targetKey = Integer.parseInt(st.nextToken());

            // ex) deque: [0, 1, 2, 3, 4, 5]
            Deque<Integer> deque = new ArrayDeque<>();
            // ex) map : {0=1, 1=1, 2=9, 3=1, 4=1, 5=1}
            HashMap<Integer, Integer> map = new HashMap<>();
            int order = 0;

            // key: 순서 번호 키, value: 중요도
            st = new StringTokenizer(br.readLine());
            int max = 0;
            for(int k = 0; k < docCount; k++){
                int importance = Integer.parseInt(st.nextToken());
                if(max < importance){
                    max = importance;
                }
                map.put(k, importance);
                deque.addLast(k);
            }

            while(!deque.isEmpty() && map.containsKey(targetKey)){
                int searchKey = deque.peekFirst();
                int peekValue = map.get(searchKey);
                if(peekValue >= max){
                    deque.pollFirst();
                    order++;
                    map.remove(searchKey);
                    if(!map.isEmpty()){ max = Collections.max(map.values()); }
                } else {
                    deque.addLast(deque.pollFirst());
                }
            }
            sb.append(order).append("\n");
        }

        System.out.print(sb);
    }
}

File: 백준/Silver/1966. 프린터 큐/프린터 큐.kt
import java.io.BufferedReader
import java.io.StringReader
import java.io.InputStreamReader
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val sb = StringBuilder()

    val loopCount = br.readLine().toInt()
    // loop scope
    for(i in 0..< loopCount) {
        var st = StringTokenizer(br.readLine());
        val docCount = st.nextToken().toInt()
        val targetKey = st.nextToken().toInt()

        // ex) deque: [{0,1}, {1, 1}, {2, 9}, {3, 1}, {4, 1} , {5, 1}]
        val deque = ArrayDeque<Pair<Int, Int>>()
        var order = 0;

        // 입력 받은 문서 갯수만큼 우선 순위 푸쉬
        st = StringTokenizer(br.readLine());
        repeat(docCount) { key ->
            val importance = st.nextToken().toInt()
            deque.addLast(key to importance)
        }

        val targetPair = deque[targetKey]
        var max = deque.maxOf { it.second }
        while(!deque.isEmpty() && deque.contains(targetPair)) {
            val pair = deque[0]
            val peekValue = pair.second
            if(peekValue >= max){
                deque.removeFirst()
                order++
                max = if(deque.isNotEmpty()) { deque.maxOf { it.second } } else { max }
            } else {
                deque.addLast(deque.removeFirst());
            }
        }
        sb.append(order).append("\n");
    }
    print(sb)
}